File: go/various/maxsubarray_test.go
package various

import "testing"

var tests = []struct {
	In                      []int
	WantLI, WantRI, WantSum int
}{
	{[]int{}, 0, 0, 0},
	{[]int{-1}, 0, 1, -1},
	{[]int{3, -1, -1, 4}, 0, 4, 5},        //whole thing
	{[]int{-1, 1, 1, -1}, 1, 3, 2},        //crossing
	{[]int{-1, -2, 1, 2}, 2, 4, 3},        //right side
	{[]int{1, 2, -3, -4}, 0, 2, 3},        //left side
	{[]int{1, -2, -3, 5, 6, 7}, 3, 6, 18}, // 6 length, right side
	{[]int{1, -2, -3, 5, 6}, 3, 5, 11},    // 5 length, right side
}

func TestMaxSubarrayRecursive(t *testing.T) {
	for _, test := range tests {
		li, ri, sum := MaxSubarrayRecursive(test.In, 0, len(test.In))
		if li != test.WantLI || ri != test.WantRI || sum != test.WantSum {
			t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for array %v", li, ri, sum, test.WantLI, test.WantRI, test.WantSum, test.In)
		}
	}
	li, ri, sum := MaxSubarrayRecursive(tests[2].In, 2, len(tests[2].In)) // test half
	if li != 3 && ri != 4 && sum != 4 {
		t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for right half test of array %v", li, ri, sum, 3, 4, 4, tests[2].In)
	}
	li, ri, sum = MaxSubarrayRecursive(tests[3].In, 0, 2) // test left
	if li != 1 && ri != 2 && sum != 1 {
		t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for right half test of array %v", li, ri, sum, 1, 2, 1, tests[3].In)
	}
	li, ri, sum = MaxSubarrayRecursive(tests[5].In, 1, 3) // test middle
	if li != 1 && ri != 2 && sum != 2 {
		t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for right half test of array %v", li, ri, sum, 1, 2, 2, tests[5].In)
	}

}

func TestMaxSubarray(t *testing.T) {
	for _, test := range tests {
		li, ri, sum := MaxSubarray(test.In, 0, len(test.In))
		if li != test.WantLI || ri != test.WantRI || sum != test.WantSum {
			t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for array %v", li, ri, sum, test.WantLI, test.WantRI, test.WantSum, test.In)
		}
	}
	li, ri, sum := MaxSubarray(tests[2].In, 2, len(tests[2].In)) // test half
	if li != 3 && ri != 4 && sum != 4 {
		t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for right half test of array %v", li, ri, sum, 3, 4, 4, tests[2].In)
	}
	li, ri, sum = MaxSubarray(tests[3].In, 0, 2) // test left
	if li != 1 && ri != 2 && sum != 1 {
		t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for right half test of array %v", li, ri, sum, 1, 2, 1, tests[3].In)
	}
	li, ri, sum = MaxSubarray(tests[5].In, 1, 3) // test middle
	if li != 1 && ri != 2 && sum != 2 {
		t.Errorf("test results (%v, %v, %v) != test expected (%v, %v, %v) for right half test of array %v", li, ri, sum, 1, 2, 2, tests[5].In)
	}
}

File: c/various/max_subarray_test.c
#include <stdio.h>

#include "max_subarray.h"

typedef struct {
  int *In;
  int InStart, InEnd, WantLI, WantRI, WantSum;
} test_vals;

int main(int argc, char **argv) {
  int failed = 0; // false

  int a0[1] = {-1};
  int a1[4] = {3,-1,-1,4}; // whole thing
  int a2[4] = {-1,1,1,-1}; // crossing middle
  int a3[4] = {-1,-2,1,2}; // right side
  int a4[4] = {1,2,-3,-4}; // left side
  int a5[6] = {1,-2,-3,5,6,7}; // 6 length, right side
  int a6[5] = {1,-2,-3,5,6}; //5 length, right side

  int test_count = 10;
  test_vals tests[10] = { // test count
    {a0, 0, 1, 0, 1, -1}, // array, start, end, want li, want ri, want sum
    {a1, 0, 4, 0, 4, 5},
    {a2, 0, 4, 1, 3, 2},
    {a3, 0, 4, 2, 4, 3},
    {a4, 0, 4, 0, 2, 3},
    {a5, 0, 6, 3, 6, 18},
    {a6, 0, 5, 3, 5, 11},
    {a6, 0, 3, 0, 1, 1},
    {a6, 3, 5, 3, 5, 11},
    {a6, 1, 3, 1, 2, -2},
  };

  for (int i = 0; i < test_count; i++) {
    max_info info = max_subarray(tests[i].In, tests[i].InStart, tests[i].InEnd);
    if (info.l != tests[i].WantLI || 
        info.r != tests[i].WantRI ||
        info.sum != tests[i].WantSum) {
      printf("failure on %d, ret info: (%d, %d, %d), expected: (%d, %d, %d)\n",
          i, info.l, info.r, info.sum, 
          tests[i].WantLI, tests[i].WantRI, tests[i].WantSum);
      failed = 1; // true
    }
  }

  if (!failed) {
    printf("all iterative subarray tests good\n");
  }

  failed = 0;

  for (int i = 0; i < test_count; i++) {
    max_info info = max_subarray_recursive(tests[i].In, tests[i].InStart, tests[i].InEnd);
    if (info.l != tests[i].WantLI || 
        info.r != tests[i].WantRI ||
        info.sum != tests[i].WantSum) {
      printf("failure on %d, ret info: (%d, %d, %d), expected: (%d, %d, %d)\n",
          i, info.l, info.r, info.sum, 
          tests[i].WantLI, tests[i].WantRI, tests[i].WantSum);
      failed = 1; // true
    }
  }

  if (!failed) {
    printf("all recursive subarray tests good\n");
  } 
}

